pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/BookkeeperSchemaStorage.java

```diff
@@ -0,0 +1,477 @@
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.KeeperException.Code;
+import org.apache.zookeeper.ZooDefs;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.data.ACL;
+
+public class BookkeeperSchemaStorage implements SchemaStorage {
+    private static final String SchemaPath = "/schemas";
+    private static final List<ACL> Acl = ZooDefs.Ids.OPEN_ACL_UNSAFE;
+
+    private final PulsarService pulsar;
+    private final ZooKeeper zooKeeper;
+    private final ZooKeeperCache localZkCache;
+    private BookKeeper bookKeeper;
+
+    @VisibleForTesting
+    BookkeeperSchemaStorage(PulsarService pulsar) {
+        this.pulsar = pulsar;
+        this.localZkCache = pulsar.getLocalZkCache();
+        this.zooKeeper = localZkCache.getZooKeeper();
+    }
+
+    @VisibleForTesting
+    public void init() throws KeeperException, InterruptedException {
+        try {
+            zooKeeper.create(SchemaPath, new byte[]{}, Acl, CreateMode.PERSISTENT);
+        } catch (KeeperException.NodeExistsException error) {
+            // race on startup, ignore.
+        }
+    }
+
+    public void start() throws IOException {
+        this.bookKeeper = pulsar.getBookKeeperClientFactory().create(
+            pulsar.getConfiguration(),
+            pulsar.getZkClient()
+        );
+    }
+
+    @Override
+    public CompletableFuture<SchemaVersion> put(String key, byte[] value, byte[] hash) {
+        return putSchemaIfAbsent(key, value, hash).thenApply(LongSchemaVersion::new);
+    }
+
+    @Override
+    public CompletableFuture<StoredSchema> get(String key, SchemaVersion version) {
+        if (version == SchemaVersion.Latest) {
+            return getSchema(key);
+        } else {
+            LongSchemaVersion longVersion = (LongSchemaVersion) version;
+            return getSchema(key, longVersion.getVersion());
+        }
+    }
+
+    @Override
+    public CompletableFuture<SchemaVersion> delete(String key) {
+        return deleteSchema(key).thenApply(LongSchemaVersion::new);
+    }
+
+    @NotNull
+    private CompletableFuture<StoredSchema> getSchema(String schemaId) {
+        return getSchemaLocator(getSchemaPath(schemaId)).thenCompose(locator -> {
+
+            if (!locator.isPresent()) {
+                return completedFuture(null);
+            }
+
+            SchemaStorageFormat.SchemaLocator schemaLocator = locator.get().locator;
+            return readSchemaEntry(schemaLocator.getInfo().getPosition())
+                .thenApply(entry ->
+                    new StoredSchema(
+                        entry.getSchemaData().toByteArray(),
+                        new LongSchemaVersion(schemaLocator.getInfo().getVersion()),
+                        emptyMap()
+                    )
+                );
+        });
+    }
+
+    @Override
+    public SchemaVersion versionFromBytes(byte[] version) {
+        ByteBuffer bb = ByteBuffer.wrap(version);
+        return new LongSchemaVersion(bb.get());
```

Comment:
get() is just reading the first byte here and ignoring the remaining 7

Reply:
Yup. Good catch.